integration/lsp.test.ts

```diff
@@ -0,0 +1,78 @@
+describe('LSP Server', () => {
+    let server;
+
+    beforeAll(async () => {
+        const { initLog } = await import('@influxdata/flux-lsp-node');
+        initLog();
+    });
+
+    beforeEach(async () => {
+        const { Lsp } = await import('@influxdata/flux-lsp-node');
+        server = new Lsp();
+    });
+
+    it('responds to initialize request', async () => {
+        const callback = jest.fn((message) => { });
+
+        const request = '{"jsonrpc": "2.0", "id": 1, "method": "initialize", "params": { "capabilities": {}}}';
+        //server.onMessage(callback);
+        const response = JSON.parse(await server.send(request));
+        expect(response).toBeDefined()
+        expect(response).toHaveProperty('result.capabilities')
+
+
+        await server.run();
+        expect(callback).not.toHaveBeenCalled();
+    });
+
+    it('throws error on bad JSON message', async () => {
+        const callback = jest.fn((message) => {
+            console.log('callback', message);
+        });
+
+        const request = 'Content-Length: 84\r\n\r\n{"jsonrpc": "2.0", "id": 1, "method": "initialize", "params": { "capabilities": {}}}';
+        server.onMessage(callback);
+        await expect(async () => {
+            try {
+                await server.send(request)
+            } catch (e) {
+                // jest can only assert on actual Errors being thrown not arbitrary expressions.
+                // Catch anything and wrap it in an Error.
+                throw new Error(e)
+            }
+        })
+            .rejects
+            .toThrow('failed to decode message JSON');
+
+        await server.run();
+        expect(callback).not.toHaveBeenCalled();
+    });
+
+    it('sends diagnostics', async () => {
+        const callback = jest.fn((message) => {
+            console.log('callback', message);
+        });
+        server.onMessage(callback);
+
+        const initRequest = '{"jsonrpc": "2.0", "id": 1, "method": "initialize", "params": { "capabilities": {}}}';
+        await server.send(initRequest)
+        const openRequest = '{"jsonrpc": "2.0", "id": 2, "method": "didOpen", "params": { "textDocument": {"text":"x ="}}}';
+        await server.send(openRequest)
+        await server.run();
```

Comment:
I think these tests time out because `server.run` never returns (or, at least, if it does, it's because an error occurred). I ran into this exact same issue and, in fact, assumed that the only way I could really fix it is to have a `server.shutdown` method that causes `server.run` to resolve its promise. I'm not sure how I feel about that, but I _assume_ that's the right thing.

Reply:
Something I've seen in this situation is to send a message to the server that acts as a "poison pill." If the callback looks for a shutdown message, and the messages are buffered in order prior to the `run()` method being invoked, the callback could invoke `shutdown()` after all the messages have been processed.

Reply:
Since JS promises are evaluated eagerly we don't need to await it though, right? Or if we want to ensure that it terminates I think it should be written as

```js
const runPromise = server.run();
expect(callback).toHaveBeenCalled();
server.send(shutdownRequest);
await server.run();
```

Reply:
It could work. The question is the timing between `server.run()` firing and the runtime yielding back to the execution of the `expect()` (which could happen before the pending messages have been processed.

I think if the `expect()` was moved to just after the `await runPromise`, that case would be covered.